polar-core/src/vm.rs

```diff
@@ -2289,41 +2289,46 @@ impl PolarVirtualMachine {
-
-                    return self.append_goals(vec![
-                        Goal::IsSubspecializer {
-                            answer: answer.clone(),
-                            left: left_spec.clone(),
-                            right: right_spec.clone(),
-                            arg: arg.clone(),
-                        },
-                        Goal::Unify {
-                            left: Term::new_temporary(Value::Variable(answer)),
-                            right: Term::new_temporary(Value::Boolean(true)),
-                        },
-                    ]);
+            match (&left_param.specializer, &right_param.specializer) {
+                (Some(left_spec), Some(right_spec)) => {
+                    // If you find two non-equal specializers, that comparison determines the relative
+                    // specificity of the two rules completely. As soon as you have two specializers
+                    // that aren't the same and you can compare them and ask which one is more specific
+                    // to the relevant argument, you're done.
+                    if left_spec != right_spec {
+                        let answer = self.kb.read().unwrap().gensym("is_subspecializer");
+                        // Bind answer to false as a starting point in case is subspecializer doesn't
+                        // bind any result.
+                        // This is done here for safety to avoid a bug where `answer` is unbound by
+                        // `IsSubspecializer` and the `Unify` Goal just assigns it to `true` instead
+                        // of checking that is is equal to `true`.
+                        self.bind(&answer, Term::new_temporary(Value::Boolean(false)));
+
+                        return self.append_goals(vec![
+                            Goal::IsSubspecializer {
+                                answer: answer.clone(),
+                                left: left_spec.clone(),
+                                right: right_spec.clone(),
+                                arg: arg.clone(),
+                            },
+                            Goal::Unify {
+                                left: Term::new_temporary(Value::Variable(answer)),
+                                right: Term::new_temporary(Value::Boolean(true)),
+                            },
+                        ]);
+                    }
                 }
+                // If the left rule has no specializer and the right does, it is NOT more specific,
+                // so we fall through to Backtrack (fail)
+                (None, Some(_)) => (),
+                // If the left rule has a specializer and the right does not, the left IS more specific,
+                // so we return
+                (Some(_), None) => return Ok(()),
+                // If neither has a specializer, neither is more specific, so we fail
+                (None, None) => (),
```

Comment:
Here, I don't think we want to fail, but just want to continue.  Right now specificity is equal, so we must go to the next argument.

Reply:
Isn't that what this does?

Reply:
Yes, in contrast to the first case though & the comment is incorrect.  We aren't failing.  We are just continuing.